Development/nmos/authorization_operation.cpp

```diff
@@ -0,0 +1,2004 @@
+                    }
+
+                }).then([&](pplx::task<void> finally)
+                {
+                    auto lock = model.write_lock(); // in order to update local state
+
+                    try
+                    {
+                        finally.get();
+
+                        registered = true;
+                    }
+                    catch (const web::http::http_exception& e)
+                    {
+                        slog::log<slog::severities::error>(gate, SLOG_FLF) << "OpenID Connect Authorization API client metadata retreieve HTTP error: " << e.what() << " [" << e.error_code() << "]";
+
+                        authorization_service_error = true;
+                    }
+                    catch (const web::json::json_exception& e)
+                    {
+                        slog::log<slog::severities::error>(gate, SLOG_FLF) << "OpenID Connect Authorization API client metadata retreieve JSON error: " << e.what();
+
+                        authorization_service_error = true;
+                    }
+                    catch (const std::exception& e)
+                    {
+                        slog::log<slog::severities::error>(gate, SLOG_FLF) << "OpenID Connect Authorization API client metadata retreieve error: " << e.what();
+
+                        authorization_service_error = true;
+                    }
+                    catch (const authorization_exception&)
+                    {
+                        slog::log<slog::severities::error>(gate, SLOG_FLF) << "OpenID Connect Authorization API client metadata retreieve error";
+
+                        authorization_service_error = true;
+                    }
+                    catch (...)
+                    {
+                        slog::log<slog::severities::severe>(gate, SLOG_FLF) << "OpenID Connect Authorization API client metadata retreieve unexpected unknown exception";
+
+                        authorization_service_error = true;
+                    }
+
+                    model.notify();
+                });
+
+                // wait for the request because interactions with the Authorization API endpoint must be sequential
+                condition.wait(lock, [&] { return shutdown || authorization_service_error || registered; });
+
+                cancellation_source.cancel();
+                // wait without the lock since it is also used by the background tasks
+                nmos::details::reverse_lock_guard<nmos::write_lock> unlock{ lock };
+                request.wait();
+                return !authorization_service_error && registered;
+            }
+
+            // register client to the Authorization server
+            bool client_registration(nmos::base_model& model, nmos::experimental::authorization_state& authorization_state, nmos::load_ca_certificates_handler load_ca_certificates, nmos::experimental::save_authorization_client_handler save_authorization_client, slog::base_gate& gate)
```

Comment:
How is this function different from `request_client_metadata_from_openid_connect`? 

They seem to have a lot of the same logic - could the common code be factored out?